switch php version

jrbarnard opened the pull request:
A currently working version of #378 which also adds some functionality that would support #708 .

Usage:
```
valet use php@7.2
valet use php7.2 // We will convert to php@7.2
valet use php72 // We will convert to php@7.2
```
All passed php versions are checked against the supported versions in the Brew class.

Couple of things this has brought up, there is a current bug which I've documented here: #709 (i'll put a PR in for it)
Valet doesn't want to work on my machine when I've linked php@5.6, I'm running high sierra and I think this is just a compatibility bug rather than with this feature, I'd value someone else having a test though.

mattstauffer commented:
I did a quick skim and it looks good, but I look forward to review from the folks who know more than me. For now I'll download and test it out--thanks for your work on this!

mattstauffer commented:
OK, I was running php 7.2 and this is what I saw:

<img width="669" alt="image" src="https://user-images.githubusercontent.com/151829/51478489-9b852580-1d59-11e9-9f1d-d54989e5468e.png">

I'll look into it more, gotta run take care of kids.

Here's my current installed list:

<img width="120" alt="image" src="https://user-images.githubusercontent.com/151829/51478530-b061b900-1d59-11e9-9eee-c941e36ebe23.png">

My guess is it hsa something to do with it being `php` instead of `php@7.2` but this whole new Homebrew versioning system is very confusing to me.

jrbarnard commented:
@mattstauffer That is confusing, I presume your local has had php@7.2 linked to php for a while, I'll have a look at a workaround that works for when php is linked but is not the latest (e.g php@7.3).

mattstauffer commented:
That was it. Since I was using `php`, not `php@7.2`, everything broke. So we need to allow for that. I installed `php@7.2` and then swapping back and forth worked fine:

<img width="685" alt="image" src="https://user-images.githubusercontent.com/151829/51478763-60372680-1d5a-11e9-95b7-669cdf549814.png">

mattstauffer commented:
@jrbarnard Yep, my guess is if someone runs `brew install php` while it's at 7.1, that still stays linked at 7.1 even when the latest becomes 7.2 or 7.3. So we need to figure out to run our uninstall based on which one is linked, whether that's `php` or `php@7.1` or `php@7.2` or whatever, rather than assuming we can construct `php@currentlyInstalledVersion`.

That's my guess.

jrbarnard commented:
@mattstauffer I've updated it to take from the linked path the php directory, e.g if latest (at any time) it should be under php, so we brew unlink php as opposed to brew unlink php@7.2.

It's using the regex from Brew::linkedPhp() so should handle the older formats too, I've updated the tests and added new ones to cover the refactor.

mattstauffer commented on cli/Valet/PhpFpm.php:
```diff
@@ -120,4 +118,53 @@ function fpmConfigPath()
+
+        if (!$this->brew->supportedPhpVersions()->contains($version)) {
+            throw new DomainException(
+                sprintf('Valet doesnn\'t support PHP version: %s', $version)
```
Typo in "doesn't"
but also maybe we should suggest one?

```php
sprintf("Valet doesn't support PHP versions: %s (try something like 'php71' instead)"
```

And that does also make me curious whether we should consider using "7.1" instead of "php7.1" or whatever as our syntax? `valet use 7.1`

mattstauffer commented:
<img width="638" alt="image" src="https://user-images.githubusercontent.com/151829/51507949-4ed53500-1dc1-11e9-8cdb-859139e55bc8.png">

I'm surprised that Homebrew makes this difficult, but I think that is a Homebrew thing, not a Valet thing. However, I still want to figure this out.

mattstauffer commented:
https://www.reddit.com/r/PHP/comments/a4vggg/php_70_removed_from_homebrew/

Hrm. I bet we're going to get some requests for that one. But I think that can be a matter of another day.

jrbarnard commented:
Hi @mattstauffer Good spot with the typo I'll update.

Syntax wise originally I set it up so you could leave off the php prefix however put it back on as 'valet use' is quite generic so wanted to be more explicit in what you're asking valet to use.

I can add a suggested PHP version, maybe if we stick with latest stable and stick to the newer 7.2 syntax? Technically use will convert the old to new anyway but probably still worth using the new.

jrbarnard commented:
Hi @mattstauffer Good spot with the typo I'll update.

Syntax wise originally I set it up so you could leave off the php prefix however put it back on as 'valet use' is quite generic so wanted to be more explicit in what you're asking valet to use.

I can add a suggested PHP version, maybe if we stick with latest stable and stick to the newer 7.2 syntax? Technically use will convert the old to new anyway but probably still worth using the new.

mattstauffer commented:
<img width="681" alt="image" src="https://user-images.githubusercontent.com/151829/51544463-fdaf5a80-1e2d-11e9-86d4-0e766e8568a6.png">

Looks better! Maybe if we drop the colon it won't line wrap?

Also... the fact that we're building this with a syntax that could potentially be used with not-PHP makes some inconsistencies clear.

```
valet use mysql
```

```
valet use php71
```

That makes sense, for sure. Although in that context, our error would be "`7.1` isn't on our list of installable packages. Here's the full list: []"

The bigger question is, are we *going to* allow for supporting other packages? My guess is "no" but I can't say for sure. Let's say we weren't. Is there a better syntax? I agree that `valet use 7.1` is too vague. Trying to think of a verb other than `use` that communicates "install PHP version" but I have nothing yet.

So.. I don't know exactly. Gonna be thinking on this a bit and I welcome feedback from others

jrbarnard commented:
The only thing I can think is we set it up like:
```
valet use php 7.1
```

It's relatively clear then that you want to use php, version 7.1, it may be a little verbose though?
```
valet use invalid
// Valet doesn't currently support that package, we currently support ['php']

valet use php 9.1
// Valet doesn't current support that version of php, why don't you try 'valet use php 7.2'
```

mattstauffer commented:
@jrbarnard good idea. let's see if we can get folks to weigh in

jalendport commented:
Meant to comment on this back when Matt asked for some folks to weigh-in. Love the proposed changes and I agree that `use` is the best option here 👌 I've actually got a little bash script that handles the php switching for you ([`switch-php`](https://github.com/jalendport/switch-php)), but happy that this is finally making it into core. Thank you for the work on this @jrbarnard! 🙌
